internal/proxy/cache_test.go

```diff
@@ -0,0 +1,498 @@
+		DefaultEnabled: true,
+	})
+	require.NotNil(t, cacheEnabled)
+}
+
+func TestSetupCache(t *testing.T) {
+	t.Run("Valid vault cache TTL shall set up and return cache successfully", func(t *testing.T) {
+		ttlArray := []time.Duration{5 * time.Minute, 1 * time.Second}
+		for _, ttl := range ttlArray {
+			cache := setupCache(config.CacheConfig{TTL: ttl})
+			require.NotNilf(t, cache, `setupCache() returns nil with env variable: %s`, ttl)
+			assert.Equal(t, ttl, cache.timeout)
+			assert.False(t, cache.enabled)
+		}
+	})
+
+	t.Run("Invalid vault cache TTL shall fail to set up and return cache", func(t *testing.T) {
+		ttlArray := []time.Duration{-2 * time.Minute, 0}
+		for _, ttl := range ttlArray {
+			cache := setupCache(config.CacheConfig{TTL: ttl})
+			require.Nil(t, cache, `setupCache() does not return nil with ttl: %s`, ttl)
+		}
+	})
+
+	t.Run("Valid vault default cache enabled shall set up and return cache successfully", func(t *testing.T) {
+		cache := setupCache(config.CacheConfig{TTL: 5 * time.Minute, DefaultEnabled: true})
+		require.NotNil(t, cache)
+		assert.True(t, cache.enabled)
+		assert.Equal(t, 5*time.Minute, cache.timeout)
+	})
+
+	t.Run("False vault default cache enabled shall result in cache.enabled=false", func(t *testing.T) {
+		cache := setupCache(config.CacheConfig{TTL: 5 * time.Minute, DefaultEnabled: false})
+		require.NotNil(t, cache)
+		assert.False(t, cache.enabled)
+	})
+}
+
+func TestGetAfterSet(t *testing.T) {
+	t.Run("normal", func(t *testing.T) {
+		cacheData := &CacheData{
+			Header:     nil,
+			Body:       []byte(fmt.Sprint(rand.Intn(100))),
+			StatusCode: http.StatusOK,
+		}
+		cache := NewCache(config.CacheConfig{TTL: 10 * time.Second})
+		cacheKey := CacheKey{
+			Token: "blue",
+			Request: &http.Request{
+				URL: &url.URL{
+					Path: "test",
+				},
+				Header: http.Header{
+					consts.AuthHeaderName: []string{"rose"},
+				},
+			},
+			RequestBody: nil,
+		}
+		cacheKeyHash, err := computeRequestID(&cacheKey)
+		require.NoError(t, err)
+		require.NotEmpty(t, cacheKeyHash)
+		cache.Set(cacheKeyHash, cacheData)
+
+		cacheDataOut, err := cache.Get(cacheKeyHash)
+		require.NoError(t, err)
+		assert.Equal(t, cacheData, cacheDataOut, `cache.Get() result doesn't match what was set with key: %s`, cacheKey)
+	})
+
+	t.Run("expired item not returned", func(t *testing.T) {
+		cacheData := &CacheData{
+			Header:     nil,
+			Body:       []byte(fmt.Sprint(rand.Intn(100))),
+			StatusCode: http.StatusOK,
+		}
+		cache := NewCache(config.CacheConfig{TTL: 1 * time.Second})
+		cacheKey := CacheKey{
+			Token: "blue",
+			Request: &http.Request{
+				URL: &url.URL{
+					Path: "test",
+				},
+				Header: http.Header{
+					consts.AuthHeaderName: []string{"rose"},
+				},
+			},
+			RequestBody: nil,
+		}
+		cacheKeyHash, err := computeRequestID(&cacheKey)
+		require.NoError(t, err)
+		require.NotEmpty(t, cacheKeyHash)
+		cache.Set(cacheKeyHash, cacheData)
+
+		time.Sleep(5 * time.Second)
```

Comment:
Can we make the expiry and sleep time something like 10ms and 20ms respectively? The cache package's own tests use times of that order of magnitude, and looking at the implementation, I don't think it will be flaky.

Reply:
Cool, changed it to 10ms and 20ms in bd90b345c483bfb20438482fa489a3ef54380bfc and it seems to be working fine.